OutOfSchool/OutOfSchool.WebApi/Services/Database/WorkshopService.cs

```diff
@@ -220,6 +306,39 @@ public async Task Delete(Guid id)
             }
         }
 
+        /// <inheritdoc/>
+        /// <exception cref="DbUpdateConcurrencyException">If a concurrency violation is encountered while saving to database.</exception>
+        /// <exception cref="InvalidOperationException">If unreal to delete images.</exception>
+        public async Task DeleteV2(Guid id)
+        {
+            logger.LogInformation($"Deleting {nameof(Workshop)} with Id = {id} started.");
+
+            var entity = await workshopRepository.GetById(id).ConfigureAwait(false);
+
+            var removingResult = await workshopImagesInteractionService.RemoveManyImagesAsync(entity.Id, entity.Images.Select(x => x.ExternalStorageId).ToList()).ConfigureAwait(false);
+
+            if (entity.Images.Count > 0 && removingResult.MultipleKeyValueOperationResult is { Succeeded: false })
+            {
+                throw new InvalidOperationException($"Unreal to delete {nameof(Workshop)} [id = {id}] because unable to delete images.");
+            }
+
+            foreach (var teacher in entity.Teachers.ToList())
+            {
+                await teacherService.Delete(teacher.Id).ConfigureAwait(false);
+            }
+
+            try
+            {
+                await workshopRepository.Delete(entity).ConfigureAwait(false);
+                logger.LogInformation($"{nameof(Workshop)} with Id = {id} successfully deleted.");
+            }
+            catch (DbUpdateConcurrencyException ex)
+            {
+                logger.LogError(ex, $"Deleting {nameof(Workshop)} with Id = {id} failed.");
+                throw;
+            }
```

Comment:
I think we need a transaction here or maybe there is a cascade deletion and will be enough to delete workshop.

Reply:
Now it will fall down if unreal to delete workshop images -> any teacher (images + data) -> workshop data so we can have a situation when a workshop isn't removed but some images/teachers are deleted (the partial delete). In this way we won't have unreferenced data in our db (garbage). We will create a transaction for these cases in the closest PRs when move our images to another storage.